src/UI/URLBuilder.php

```
@@ -0,0 +1,281 @@
+     * changed URLBuilder as well as the token for any
+     * subsequent changes to the acquired parameter.
+     *
+     * @return array<URLBuilder,URLBuilderToken>
+     * @throws \ilException
+     * @throws \InvalidArgumentException
+     */
+    public function acquireParameter(array $namespace, string $name, ?string $initial_value = null): array
+    {
+        if ($name === '' || empty($namespace)) {
+            throw new \InvalidArgumentException("Parameter name or namespace not set");
+        }
+
+        $parameter = implode(self::SEPARATOR, $namespace) . self::SEPARATOR . $name;
+        if ($this->parameterExists($parameter)) {
+            throw new \ilException("Parameter '" . $parameter . "' already reserved in URL");
+        }
+
+        $token = new URLBuilderToken($namespace, $name);
+        $clone = clone $this;
+        $clone->params[$parameter] = ($initial_value) ?? '';
+        $clone->tokens[$parameter] = $token;
+
+        return [$clone, $token];
+    }
+
+    public function acquireParameters(array $namespace, string ...$names): array
+    {
+        $tokens = [];
+        $builder = $this;
+        foreach ($names as $name) {
+            list($builder, $token) = $builder->acquireParameter($namespace, $name);
+            $tokens[] = $token;
+        }
+        array_unshift($tokens, $builder);
+        return $tokens;
+    }
+
+    /**
+     * Delete an acquired parameter if the supplied token is valid
+     */
+    public function deleteParameter(URLBuilderToken $token): self
+    {
+        $this->checkToken($token);
+        $clone = clone $this;
+        unset($clone->params[$token->getName()]);
+        unset($clone->tokens[$token->getName()]);
+
+        return $clone;
+    }
+
+    /**
+     * Change an acquired parameter's value if the supplied token is valid
+     */
+    public function withParameter(URLBuilderToken $token, $value): self
+    {
+        assert(ist_string($value) || is_array($value));
```

I think this line has been added for backwards compatibility, right? Please don't use `assert()` though, and throw an `InvalidArgumentException` yourself instead. Also note the typo in `ist_string()`.